tango_simlib/utilities/validate_device.py

```diff
@@ -0,0 +1,415 @@
+            tango_device_data["meta"]["properties"],
+            bidirectional,
+        )
+    )
+
+    return "\n".join(issues)
+
+
+def check_list_dict_differences(spec_data, dev_data, type_str, bidirectional):
+    """Compare Commands and Attributes in the parsed YAML
+
+    Parameters
+    ----------
+    spec_data : list
+        List of dictionaries with specification data
+        E.g [
+                {'disp_level': 'OPERATOR',
+                 'doc_in': 'ON/OFF',
+                 'doc_out': 'Uninitialised',
+                 'dtype_in': 'DevBoolean',
+                 'dtype_out': 'DevVoid',
+                 'name': 'Capture'},
+                 ...
+            ]
+
+    dev_data : list
+        List of dictionaries with device data
+        E.g [
+                {'disp_level': 'OPERATOR',
+                 'doc_in': 'ON/OFF',
+                 'doc_out': 'Uninitialised',
+                 'dtype_in': 'DevBoolean',
+                 'dtype_out': 'DevVoid',
+                 'name': 'Capture'},
+                 ...
+            ]
+
+    type_str : str
+        Either "Command" or "Attribute"
+
+    bidirectional: bool
+        Whether to include details on the device that is not in the specification
+
+    Returns
+    -------
+    list
+        A list of strings describing the issues, empty list for no issues
+    """
+    issues = []
+    spec_data = sorted(spec_data, key=lambda i: i["name"])
+    dev_data = sorted(dev_data, key=lambda i: i["name"])
+
+    # Check that the command/attribute names match
+    spec_data_names = {i["name"] for i in spec_data}
+    dev_data_names = {i["name"] for i in dev_data}
+    if spec_data_names != dev_data_names:
+        diff = spec_data_names.difference(dev_data_names)
+        if diff:
+            diff = sorted(diff)
+            issues.append(
+                "{} differs, [{}] specified but missing in device".format(
+                    type_str, ",".join(diff)
+                )
+            )
+
+        if bidirectional:
+            diff = dev_data_names.difference(spec_data_names)
+            if diff:
+                diff = sorted(diff)
+                issues.append(
+                    "{} differs, [{}] present in device but not specified".format(
+                        type_str, ",".join(diff)
+                    )
+                )
+
+    # Check that the commands/attributes (by name) that are in both the spec and device
+    # are the same
+    mutual_names = spec_data_names.intersection(dev_data_names)
+    mutual_spec_data = filter(lambda x: x["name"] in mutual_names, spec_data)
+    mutual_dev_data = filter(lambda x: x["name"] in mutual_names, dev_data)
+    for spec, dev in zip(mutual_spec_data, mutual_dev_data):
+        issues.extend(check_single_dict_differences(spec, dev, type_str, bidirectional))
+
+    return issues
+
+
+def check_single_dict_differences(spec, dev, type_str, bidirectional):
+    """Compare a single attribute/property
```

Comment:
I see `type_str`can either be 'command' or 'attribute'. Please update this to match with that.